botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+        Example: Given the input array ["foo", "bar"]
+        [0] => "foo"
+
+        :type value: dict or list
+        :param value: A JSON like object that contains values, key-value pairs or
+        some combination of both.
+
+        :type path: str
+        :param path: A string representing a key, index or some combination of both.
+        A period `.` indicates the value to the right of it is a nested key in a
+        dictionary, and an integer in brackets such as `[1]` represents an index
+        of a list. These can be used together for example `Thing1.foo[3]`. An index
+        must always be at the end of a path.
+
+        :returns: The object contained within the provided path.
+        """
+        for part in path.split("."):
+            match = GET_ATTR_RE.search(part)
+            if match is not None:
+                name, index = match.groups()
+                index = int(index)
+                value = value.get(name)
+                if value is None or index >= len(value):
+                    return None
+                # the index always occurs at the end of a path
+                return value[index]
+            else:
+                value = value[part]
+        return value
+
+    def _format_partition_output(self, partition):
+        output = partition["outputs"]
+        output["name"] = partition["id"]
+        return output
+
+    def _is_partition_match(self, region, partition):
+        return (
+            region in partition["regions"]
+            or re.match(partition["regionRegex"], region) is not None
+        )
+
+    def _aws_partition(self, value):
+        """Evaluates a single string argument value as a region, and matches
+        the string value to an AWS partition. For forwards compatibility, check
+        the `regionRegex` if there is no match. If the input doesn't match any
+        region or regionRegex, return the `aws` partition, considered to be the
+        default.
+
+        :type value: str
+        :param value: A string representing an AWS region.
+
+        :returns: `output` object in the matched partition such as
+        {
+             "name": "aws",
+             "dnsSuffix": "amazonaws.com",
+             "dualStackDnsSuffix": "api.aws",
+             "supportsFIPS": true,
+             "supportsDualStack": true
+         }
+        """
+        if value is None:
+            return None
+
+        # `aws` partition is first in the partitions.json file
+        aws_partition = self.partitions["partitions"][0]
+        aws_partition_output = self._format_partition_output(aws_partition)
+        if self._is_partition_match(value, aws_partition):
+            return aws_partition_output
+
+        remaining_partitions = self.partitions["partitions"][1:]
+        for partition in remaining_partitions:
+            if self._is_partition_match(value, partition):
+                return self._format_partition_output(partition)
+
+        return aws_partition_output
```

Comment:
Out of curiousity, what do you mean by access them once? I think in the original code you're creating two copies of the partitions list because you're using a slice.